pylint/pyreverse/diagrams.py

```diff
@@ -178,24 +178,41 @@ def extract_relationships(self):
                 except KeyError:
                     continue
             # implements link
-            for impl_node in node.implements:
-                try:
-                    impl_obj = self.object_from_node(impl_node)
-                    self.add_relationship(obj, impl_obj, "implements")
-                except KeyError:
-                    continue
+            if hasattr(node, "implements"):
+                for impl_node in node.implements:
+                    try:
+                        impl_obj = self.object_from_node(impl_node)
+                        self.add_relationship(obj, impl_obj, "implements")
+                    except KeyError:
+                        continue
             # associations link
             for name, values in list(node.instance_attrs_type.items()) + list(
                 node.locals_type.items()
             ):
                 for value in values:
+                    association_type = "association"
                     if value is astroid.Uninferable:
                         continue
+
+                    # Handle multiplicity link
+                    if isinstance(value, astroid.nodes.node_classes.Subscript):
+                        children = list(value.get_children())
+                        if (
+                            len(children) == 2
+                            and children[0].name == "list"
+                            and list(children[1].infer())
+                            and list(children[1].infer())[0] is not astroid.Uninferable
+                        ):
```

Comment:
This currently only handles the very special case that the attribute is annotated with `list[<type>]`.
I think it would be a good idea to extract two utility functions:
* One that replaces the `isinstance(value, astroid.nodes.node_classes.Subscript)` and determines if this is a multiplicity link or not
* One that replaces the body of the `if` block, and infers the `value`

This would make the code more readable, especially as this section will probably get longer if we also want to handle things like: 
```python
class Dummy: pass
class MyClass:
    def __init__(self):
        self.multiplicity_1: tuple[Dummy] = (Dummy(), Dummy())  # type hinted, but with tuple
        self.mulitplicity_2: Iterable[Dummy] = [Dummy(), Dummy()]  # type hinted, but with generic Iterable
        self.mulitplicity_3 = [Dummy(), Dummy()]  # no type annotations to rely on
```

By the way, if you want to rebase on the current main branch, I have implemented an easier and more flexible way for functional tests with `pyreverse` in #6449. That way you don't have to include all those examples in the existing test data under `tests/data` and subsequently update all output formats under `pyreverse/ data`.